SlicerSofa/SlicerSofa.py

```diff
@@ -19,148 +51,490 @@
 
 from SofaEnvironment import *
 
-#
-# SlicerSofa
-#
+# -----------------------------------------------------------------------------
+# Decorator: RunOnce
+# -----------------------------------------------------------------------------
+def RunOnce(func):
+    """
+    Decorator that marks a function to be executed only once.
 
+    Attributes:
+        runOnce (bool): Flag indicating the function should run only once.
+    """
+    func.runOnce = True
+    return func
 
-class SlicerSofa(ScriptedLoadableModule):
-    """Uses ScriptedLoadableModule base class, available at:
-    https://github.com/Slicer/Slicer/blob/main/Base/Python/slicer/ScriptedLoadableModule.py
+# -----------------------------------------------------------------------------
+# Utility Functions
+# -----------------------------------------------------------------------------
+def arrayFromMarkupsROIPoints(roiNode):
+    """
+    Utility function to return RAS (Right-Anterior-Superior) boundaries from a vtkMRMLMarkupsROINode.
+
+    Args:
+        roiNode (vtkMRMLMarkupsROINode): The ROI node from which to extract boundaries.
+
+    Returns:
+        list: A list containing [R_min, A_min, S_min, R_max, A_max, S_max].
+              Returns [0.0, 0.0, 0.0, 0.0, 0.0, 0.0] if roiNode is None.
     """
+    if roiNode is None:
+        return [0.0] * 6
+
+    center = [0] * 3
+    roiNode.GetCenter(center)
+    size = roiNode.GetSize()
+
+    # Calculate min and max RAS bounds
+    R_min = center[0] - size[0] / 2
+    R_max = center[0] + size[0] / 2
+    A_min = center[1] - size[1] / 2
+    A_max = center[1] + size[1] / 2
+    S_min = center[2] - size[2] / 2
+    S_max = center[2] + size[2] / 2
+
+    return [R_min, A_min, S_min, R_max, A_max, S_max]
 
+def arrayVectorFromMarkupsLinePoints(lineNode):
+    """
+    Utility function to return the vector from a vtkMRMLMarkupsLineNode.
+
+    Args:
+        lineNode (vtkMRMLMarkupsLineNode): The line node from which to extract the vector.
+
+    Returns:
+        list: A list containing the vector components [x, y, z].
+              Returns [0.0, 0.0, 0.0] if lineNode is None.
+    """
+    if lineNode is None:
+        return [0.0] * 3
+
+    # Calculate vector direction and normalize
+    p1 = lineNode.GetNthControlPointPosition(0)
+    p2 = lineNode.GetNthControlPointPosition(1)
+    return [p2[0] - p1[0], p2[1] - p1[1], p2[2] - p1[2]]
```

Comment:
Maybe better to get points with `arrayFromMarkupsControlPoints` then subtract with numpy.